### ISSUE TITLE
Merge improvements (pitch bend, <s>`parseSong()`</s> song parsing, ...) from my private repo

### ISSUE THREAD

**Issue by page200** [nfroidure/MIDIFile#34]
There are various improvements in a copy of MIDIFile.js in https://github.com/surikov/webaudiofont/blame/master/examples/MIDIFile.js such as support for pitch bends (`slides`).

It might be good to merge those improvements into this repo, and replace MIDIFile.js in that repo by a link to this repo, as discussed at https://github.com/nfroidure/midifile/issues/33#issuecomment-1655372411

**Edit 2025-03-01:** Don't merge the current version of `webaudiofont`. Merge my private version, see https://github.com/nfroidure/midifile/issues/34#issuecomment-2692426531 below.

**Comment by mk-pmb** [comment]
Unfortunately they seem to use a stricter license (GNU GPL v3) than we do (MIT), so we cannot take code from their repo. We'd have to track down the relevant patch author(s) and request explicit permission from them.

Most of the code lines that have the word "slide" seem to come from [this commit](https://github.com/surikov/webaudiofont/commit/d3d29592ce29d9aeb673515b9723e272b39ad7da) by user @surikov but I cannot find a PR for that. Also the commit title "test" looks like the patch might not be a single feature, so we'd need to figure out whether we need other patches as well to support slides.

If your patch from https://github.com/surikov/webaudiofont/pull/95 works without slides, feel free to submit it as PR here under MIT license. (Assuming you do have all required copyrights, of course.)

**Comment by page200** [comment]
@surikov's patches introduce pitch bend etc., and my patch https://github.com/surikov/webaudiofont/pull/95 adds support for MIDI files with custom pitch-bend range, so that the pitch bend is correct for them.

What form of permission is needed, as comments on GitHub? You have my permission, and I think @surikov also suggests [here](https://github.com/surikov/webaudiofont/issues/93#issuecomment-1427104183) that @nfroidure's repos should be modified accordingly (and that @surikov's patches to MIDIFile.js aren't the focus of his repo anyway, but are merely a small example).

**Comment by mk-pmb** [comment]
> What form of permission is needed, as comments on GitHub?

_Obligatory disclaimer: I'm not a lawyer, and the following is just how I as a layman understand (US) copyright and (European) authorship protection._

We need an MIT license grant. A comment stating the grant would work in theory, but we'd have to keep that around somehow. The easier way is to submit a PR for a branch that is

* based on our master branch and
* whose new commits do not modify any license or authorship statements in existing files.
  * <small>(and do not introduce any new terms that contradict the current license.)</small>

The 2nd condition acts as implied license grant under the existing license. It's only effective if you're legally capable of granting the license, but this is most usually the case for your own creations.

**Comment by mk-pmb** [comment]
I checked https://github.com/surikov/webaudiofont/pull/95 and it looks like [the patch there](https://github.com/surikov/webaudiofont/commit/64ffca58d26d9ca78e433eed562d9268ed7bee9e.patch) isn't directly compatible, even if I adjust the file paths. For example, it tries to modify a function `addSlide` but that code seems to not exist on our master branch. My quick attempt to trace where `addSlide` comes from, failed. I might try again later.

Edit: If with

> surikov's patches introduce pitch bend etc.,

you mean he is the author of the patches for those features, we'd have to ask him to please PR his changes into our repo. However, they may require even more patches from other authors, can't currently check that.

**Comment by page200** [comment]
My patch fixes the pitch-bend range. @nfroidure's repos don't have pitch bend yet (as far as I know / in the form we're discussing). @surikov's repo does.

Here's the blame history for `addSlide`: https://github.com/surikov/webaudiofont/blame/8d3bbebcb1b3fa5642b931bf1eff9e7740795c51/examples/MIDIFile.js#L1009-L1025

The contributors (@surikov and me) are listed on the left side of the code and on the top right.

The code was added in subsequent commits _after_ "forking" the original MIDIFile.js.

**Comment by mk-pmb** [comment]
Thanks! So it seems the other code all comes from the ["test" commit](https://github.com/surikov/webaudiofont/commit/d3d29592ce29d9aeb673515b9723e272b39ad7da). Back then, the repo seems to have been under MIT license still, so we might be in luck. [I've started negotiations.](https://github.com/surikov/webaudiofont/issues/97)

Once we have proper permissions on that code, we'll have to figure out stuff like what this magic number means:

> if (track.notes[i].duration == 0.0000001 

and what conparison operator to use instead, probably `<=` or `===`.

Edit: Also there is a suspicious change from binary `|` to boolean `||` near the `/127` division.

**Comment by mk-pmb** [comment]
The issue where I [asked how to credit him](https://github.com/surikov/webaudiofont/issues/97) was closed without comment, so I guess he doesn't care. My approach would be to use the "From:" header of [the "test" commit as a patch file](https://github.com/surikov/webaudiofont/commit/d3d29592ce29d9aeb673515b9723e272b39ad7da.patch) as the author attribution. Would you other maintainers agree?

**Comment by page200** [comment]
Yes, that sounds like an appropriate author attribution. And feel free to attribute me for the pitch-bend range functionality, if that makes things easier. I welcome a license that benefits the project and its widespread use. Thanks!

**Comment by mk-pmb** [comment]
<del>

> And feel free to attribute me for the pitch-bend range functionality,

Are you refering to parts of the "test" commit? Are you the real author of part of that? If so, can you PR those parts first?

</del>

Edit: I see now that you probably meant the git-blame you linked above:

> Here's the blame history for addSlide: […]

I'll take that into account.

**Comment by page200** [comment]
I meant https://github.com/surikov/webaudiofont/pull/95.

> I'll take that into account.

Whatever benefits a widespread use of the project.

Thanks!

**Comment by samuelbradshaw** [comment]
Hi! I'm also interested in seeing these changes merged in. @mk-pmb, were you planning to open a pull request?

It looks like surikov/webaudiofont switched to GPL on January 16, 2020. Previous to that, it was under an MIT license.

**Comment by page200** [comment]
I have considerably improved that code in a private codebase. One of the improvements is that it now works also when the several messages for setting the pitch-bend range are interspersed with unrelated messages, which is allowed by the MIDI specification and is the case in some MIDI files. That actually made the code simpler and more consistent with the original `midifile` code. There are also other improvements.

I can try to clean up the code (remove experimental features) earlier than I had planned if someone wants it (and is willing to then help merge it?).

That seems better than spending time on merging the old version. Not only because the old version is superseded, but also because removing that old code will be tedious. I'm renaming the issue accordingly.

**Comment by samuelbradshaw** [comment]
That's great to hear! I found a solution for my immediate need, so no rush.

**Comment by mk-pmb** [comment]
Due to lack of time, I have no progress on recovering the surikov code. So if your private repo version has code that can solve the same problem, and has clean heritage, I'd love to use that instead.

**Comment by nfroidure** [comment]
Switching this repo to GNU GPL v3 is also an option for me.

**Comment by mk-pmb** [comment]
> Switching this repo to GNU GPL v3 is also an option for me.

Please don't. I have projects that aren't copyleft enough for GPL.

Let's rather hope that page200 will find time to cleanup their new and improved version.

**Comment by page200** [comment]
Maybe we can fix SpessaSynth as I described in https://github.com/spessasus/spessasynth_lib/issues/7 and use it?

Or we could use [`wavelet`](https://github.com/ryohey/wavelet). Is there a demo page, does it work well, what are the consequences of `wavelet`'s two issues about not using the main thread? I posted my questions at https://github.com/ryohey/wavelet/issues/11#issuecomment-2583038103

Either of these options would also make it easier for me to fix the present issue (because I could then omit some `webaudiofont`-related parts of my code instead of polishing them).

**Comment by mk-pmb** [comment]
> because I could then omit some webaudiofont-related parts of my code instead of polishing them

What parts where? Maybe they can be uncoupled some other way.
